instrumentation/opentelemetry-annotations-1.0/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/otelannotations/async/MethodSpanStrategies.java

```diff
@@ -0,0 +1,44 @@
+/*
+ * Copyright The OpenTelemetry Authors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package io.opentelemetry.javaagent.instrumentation.otelannotations.async;
+
+import java.util.HashMap;
+import java.util.Map;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.CompletionStage;
+
+/**
+ * Registry of {@link MethodSpanStrategy} implementations for tracing the asynchronous operations
+ * represented by the return type of a traced method.
+ */
+public class MethodSpanStrategies {
+  private static final MethodSpanStrategies instance;
+
+  static {
+    Map<Class<?>, MethodSpanStrategy> strategies = new HashMap<>();
+    strategies.put(CompletionStage.class, MethodSpanStrategy.forCompletionStage());
+    strategies.put(CompletableFuture.class, MethodSpanStrategy.forCompletableFuture());
+    instance = new MethodSpanStrategies(strategies);
+  }
+
+  public static MethodSpanStrategies getInstance() {
+    return instance;
+  }
+
+  private final Map<Class<?>, MethodSpanStrategy> strategies;
+
+  private MethodSpanStrategies(Map<Class<?>, MethodSpanStrategy> strategies) {
+    this.strategies = strategies;
+  }
+
+  public void registerStrategy(Class<?> returnType, MethodSpanStrategy strategy) {
+    strategies.put(returnType, strategy);
+  }
+
+  public MethodSpanStrategy resolveStrategy(Class<?> returnType) {
+    return strategies.getOrDefault(returnType, MethodSpanStrategy.synchronous());
```

Comment:
Hmm, this won't work if you pass some `CustomFuture` that implements `CompletionStage`: you'd have to iterate over the map and check if `entry.getKey().isAssignableFrom(returnType)`

Reply:
Although even with `CompletionStage`/`CompletableFuture` if you don't return the composed future that will cause any composition done by the consumer of the annotated method to count towards the duration of the span since the registered notifications are fired in reverse order and on the same thread by default.